src/tls/openssl.c

```diff
@@ -262,8 +262,10 @@ static void *tls_context_create(int verify, int debug,
 
         /* Make sure the key and certificate file match */
         if (SSL_CTX_check_private_key(ssl_ctx) != 1) {
-            flb_error("[tls] private_key '%s' and password don't match",
-                      key_file);
+            flb_error("[tls] private_key '%s' %lu: %s",
+                      key_file,
+                      ERR_get_error(),
```

Comment:
`ERR_get_error()` will pop the value from the error queue. The value passed to `ERR_error_string` is not the same as the value returned from the first `ERR_get_error()` call.

Reply:
Nice catch, @krispraws. To be honest, I had not checked [the documentation](https://www.openssl.org/docs/manmaster/man3/ERR_get_error.html) to understand the proper usage of `ERR_get_error()`; instead, I simply followed what the contributors have done on lines [217-218](https://github.com/fluent/fluent-bit/blob/master/src/tls/openssl.c#L217), [227-228](https://github.com/fluent/fluent-bit/blob/master/src/tls/openssl.c#L227), [242-243](https://github.com/fluent/fluent-bit/blob/master/src/tls/openssl.c#L242) and [259-260](https://github.com/fluent/fluent-bit/blob/master/src/tls/openssl.c#L259).

I can definitely change my contribution, and I believe that I should replace `ERR_get_error` with `ERR_peek_last_error` on both lines, right? As far as I could understand from the documentation, `ERR_get_error` and `ERR_peek_error` will return me the earliest/eldest error, which is not necessarily what I am looking for, either.

If the above statement is correct, do you think I should keep the MR as is, and open a new issue and an associated MR to address all the `ERR_get_error` bugs inside the function? Or, should I simply update my patch, first, and then open an issue accordingly?

Reply:
Or, assuming that there are always proper error checks inside the code, which always pops the error, I can do the following:

```c
flb_error("[tls] private_key '%s' %lu: %s",
            key_file,
            ERR_peek_error(),
            ERR_error_string(ERR_get_error(), NULL));
```

Reply:
@aytekinar, I am not an expert on OpenSSL but my understanding of their docs is that a single OpenSSL call could theoretically add multiple errors to the queue. And the oldest error would actually be the most important one since it happens first/is the root cause.
You could use `ERR_peek_error()` or just store the value from `ERR_get_error()` to use it more than once.
Ideally, the entire error queue should be printed out/or cleared after each OpenSSL call but it may be overkill in most cases. We also need a `ERR_clear_error()` call at the beginning of the `tls_context_create` so that there are no remnants in the queue.

Reply:
I agree @krispraws 's comment.

1. Call `ERR_clear_error()` to ensure no errors at the beginning of the `tls_context_create`.
2. Store a return value of `ERR_get_error()` and use it for flb_error.


e.g.
```c
unsigned long err;
(snip)
err = ERR_get_error();
flb_error("[tls] ca_path'%s' %lu: %s",
                      ca_path,
                      err,
                      ERR_error_string(err, NULL));
```

Reply:
Sure. I will check and finalize everything by the end of this weekend (hopefully, tonight).

Thank you, @krispraws and @nokute78, for your feedback!

Reply:
@aytekinar I just stumbled upon this and noticed that no one pointed out that due to how parameters are placed when a call is made your previous code would actually pop the last error code from the error stack and then peek. I can't take the time to verify the standard to see how this behavior is qualified (ie. undefined, implementation specific or something else) but I can provide an example that shows that even today with the current calling convention this behavior stands :

[godbolt C example with machine code](https://godbolt.org/z/dY3decMPK)
[calling convention reference page](https://en.wikipedia.org/wiki/X86_calling_conventions)

Again, I'm writing this just to share useful information, I understand that this might not be relevant to the current code in the PR (which I didn't review).